src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -53,66 +53,226 @@ static ucs_status_t uct_rdmacm_cm_query(uct_cm_h cm, uct_cm_attr_t *cm_attr)
         return;
     }
 
-    conn_param.private_data_len = sizeof(*hdr) + hdr->length;
+    remote_data.field_mask            = UCT_CM_REMOTE_DATA_FIELD_DEV_ADDR        |
+                                        UCT_CM_REMOTE_DATA_FIELD_DEV_ADDR_LENGTH |
+                                        UCT_CM_REMOTE_DATA_FIELD_CONN_PRIV_DATA  |
+                                        UCT_CM_REMOTE_DATA_FIELD_CONN_PRIV_DATA_LENGTH;
+    remote_data.dev_addr              = dev_addr;
+    remote_data.dev_addr_length       = addr_length;
+    remote_data.conn_priv_data        = hdr + 1;
+    remote_data.conn_priv_data_length = hdr->length;
+
+    listener->conn_request_cb(&listener->super, listener->user_data,
+                              dev_name, event, &remote_data);
+    ucs_free(dev_addr);
+}
 
-    if (rdma_connect(cep->id, &conn_param)) {
-        ucs_error("rdma_connect(to addr=%s) failed: %m",
+static void uct_rdmacm_cm_handle_event_connect_response(struct rdma_cm_event *event)
+{
+    struct sockaddr            *remote_addr = rdma_get_peer_addr(event->id);
+    uct_rdmacm_priv_data_hdr_t *hdr;
+    uct_rdmacm_cm_ep_t         *cep;
+    char                       ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    uct_device_addr_t          *dev_addr;
+    size_t                     addr_length;
+    uct_cm_remote_data_t       remote_data;
+    ucs_status_t               status;
+
+    cep = event->id->context;
+    ucs_assert(event->id == cep->id);
+
+    hdr = (uct_rdmacm_priv_data_hdr_t *)event->param.conn.private_data;
+
+    remote_data.field_mask            = UCT_CM_REMOTE_DATA_FIELD_CONN_PRIV_DATA |
+                                        UCT_CM_REMOTE_DATA_FIELD_CONN_PRIV_DATA_LENGTH;
+    remote_data.conn_priv_data        = hdr + 1;
+    remote_data.conn_priv_data_length = hdr->length;
+
+    status = uct_rdmacm_cm_id_to_dev_addr(event->id, &dev_addr, &addr_length);
+    if (status != UCS_OK) {
+        ucs_error("client (ep=%p id=%p) failed to process a connect response "
+                  "from server %s.", cep, event->id,
                   ucs_sockaddr_str(remote_addr, ip_port_str,
                                    UCS_SOCKADDR_STRING_LEN));
+        uct_rdmacm_cm_ep_client_connect_cb(cep, &remote_data, status);
+        return;
+    }
+
+    remote_data.field_mask       |= UCT_CM_REMOTE_DATA_FIELD_DEV_ADDR |
+                                    UCT_CM_REMOTE_DATA_FIELD_DEV_ADDR_LENGTH;
+    remote_data.dev_addr          = dev_addr;
+    remote_data.dev_addr_length   = addr_length;
 
+    uct_rdmacm_cm_ep_client_connect_cb(cep, &remote_data, hdr->status);
+
+    ucs_free(dev_addr);
+
+    if (rdma_establish(event->id)) {
+        ucs_error("rdma_establish on ep %p (to server addr=%s) failed: %m",
+                  cep, ucs_sockaddr_str(remote_addr, ip_port_str,
+                                        UCS_SOCKADDR_STRING_LEN));
+        uct_rdmacm_cm_ep_client_connect_cb(cep, &remote_data, UCS_ERR_IO_ERROR);
```

Comment:
do we really need to call connect callback again? it was already called before
i would expect to move ep to error state

Reply:
uct_rdmacm_cm_ep does not have state for now, IMO better to move `rdma_establish` before CB invoke and pass valid status there.

Reply:
it's not possible because then the server can try to send to RC QP which is not yet connected
the whole point of rdma_establish is to call it **after** the client handled the response

Reply:
makes sense

Reply:
then here should be "connected_cb" with error status

Reply:
@yosefe so i'm keeping it as is, yes?

Reply:
let's call disconnect callback here
since user (UCP) would think it's already connected

Reply:
Why? connect_cb with error status does not mean successful connection. IMO it's more clear to detect that connection process was failed. disconnect_cb should initiate backward disconnect process in close protocol.

Reply:
but at this point, connect_cb was already called with success